### PR TITLE
Handle errors gracefully when requiring gems

### LINKED ISSUE #3960: NameError during Runtime#require raises unexpected error
```
$ bundle -v
Bundler version 1.10.6

$ bundle console
--- ERROR REPORT TEMPLATE -------------------------------------------------------
- What did you do?

  I ran the command `/Users/andre/.gem/ruby/2.2.2/bin/bundle console`

- What did you expect to happen?

  I expected Bundler to...

- What happened instead?

  Instead, what actually happened was...


Error details

    NameError: uninitialized constant Postgres
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:76:in `const_get'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:76:in `require'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:76:in `block (2 levels) in require'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:72:in `each'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:72:in `block in require'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:61:in `each'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/runtime.rb:61:in `require'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler.rb:134:in `require'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/cli/console.rb:10:in `run'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/cli.rb:308:in `console'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/vendor/thor/lib/thor.rb:359:in `dispatch'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/vendor/thor/lib/thor/base.rb:440:in `start'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/cli.rb:10:in `start'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/bin/bundle:20:in `block in <top (required)>'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/lib/bundler/friendly_errors.rb:7:in `with_friendly_errors'
      /Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/bin/bundle:18:in `<top (required)>'
      /Users/andre/.gem/ruby/2.2.2/bin/bundle:23:in `load'
      /Users/andre/.gem/ruby/2.2.2/bin/bundle:23:in `<main>'

Environment

    Bundler   1.10.6
    Rubygems  2.4.5
    Ruby      2.2.2p95 (2015-04-13 revision 50295) [x86_64-darwin14]
    GEM_HOME  /Users/andre/src/bundler/bundler-api/.bundle/ruby/2.2.0
    GEM_PATH
    Git       2.3.0

Bundler settings

    path
      Set for your local app (/Users/andre/src/bundler/bundler-api/.bundle/config): ".bundle"
    disable_shared_gems
      Set for your local app (/Users/andre/src/bundler/bundler-api/.bundle/config): "1"
    jobs
      Set via BUNDLE_JOBS: "8"
    bin_path
      Set via BUNDLE_BIN_PATH: "/Users/andre/.gem/ruby/2.2.2/gems/bundler-1.10.6/bin/bundle"
    gemfile
      Set via BUNDLE_GEMFILE: "/Users/andre/src/bundler/bundler-api/Gemfile"
    orig_manpath
      Set via BUNDLE_ORIG_MANPATH: "/usr/local/share/man:/usr/share/man:/usr/local/share/man:/opt/X11/share/man:/usr/local/MacGPG2/share/man"
--- TEMPLATE END ----------------------------------------------------------------

Unfortunately, an unexpected error occurred, and Bundler cannot continue.

First, try this link to see if there are any existing issue reports for this error:
https://github.com/bundler/bundler/search?q=uninitialized+constant+Postgres&type=Issues

If there aren't any reports for this error yet, please create copy and paste the report template above into a new issue. Don't forget to anonymize any private data! The new issue form is located at:
https://github.com/bundler/bundler/issues/new
```

### PR DESCRIPTION
**Pull request by agis-**
Fixes #3960.

### CODE REVIEW

**Comment by agis-** [comment]
@indirect There are various options here. We could either show the complete backtrace (like this PR does), or filter it out and show only the non-bundler parts.

Also, I'm not very sure about the wording. Perhaps we want to provide a more meaningful message like

> Error while loading sinatra.

What do you think?

File: lib/bundler/friendly_errors.rb [hunk 1]
```diff
@@ -21,6 +21,10 @@ def self.with_friendly_errors
   rescue Thor::Error => e
     Bundler.ui.error e.message
     exit 1
+  rescue NameError => e
```
Comment (segiddins) [inline lib/bundler/friendly_errors.rb]: this isn't necessarily true, there could be a bug in bundler that causes name errors

Comment (agis-) [inline lib/bundler/friendly_errors.rb]: Hm true. Then we could raise a custom exception if the `#require` fails and rescue it here. This way we'll have more information like the gem name, enabling to show more helpful messages.

Comment (segiddins) [inline lib/bundler/friendly_errors.rb]: Yeah, I'd prefer to do this explicitly inside `require`

Comment (agis-) [inline lib/bundler/friendly_errors.rb]: Updated! :)

File: spec/runtime/require_spec.rb [hunk 2]
```diff
@@ -98,6 +98,20 @@
...
+    build_lib "faulty", "1.0.0" do |s|
+      s.write "lib/faulty.rb", "HI"
+    end
+
+    gemfile <<-G
+      path "#{lib_path}"
+      gem "faulty"
+    G
+
+    run "Bundler.require", expect_err: true
```
Comment (indirect) [inline spec/runtime/require_spec.rb]: you'll need to change this to `"expect_err" => true`, because we still have to support Ruby 1.8.7. :fearful:

**Comment by indirect** [comment]
This looks good to me! :green_heart: :arrow_right: :ship:

File: lib/bundler/runtime.rb [hunk 3]
```diff
@@ -73,7 +73,12 @@ def require(*groups)
             # Allow `require: true` as an alias for `require: <name>`
             file = dep.name if file == true
             required_file = file
-            Kernel.require file
+            begin
+              Kernel.require file
+            rescue => e
+              raise Bundler::GemRequireError.new e,
```
Comment (segiddins) [inline lib/bundler/runtime.rb]: Actually... Won't this mean we never get down to our loaderror handling code?

Comment (agis-) [inline lib/bundler/runtime.rb]: Oops, nice catch! Fixed.

File: spec/runtime/require_spec.rb (hunk 2 above)
Comment (agis-) [inline spec/runtime/require_spec.rb]: Oh, right. Fixed.

**Comment by segiddins** [comment]
:+1:

**Pull request merged**